fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanBalanceService.java

```diff
@@ -93,20 +102,32 @@ public boolean isOverPaid(final Loan loan) {
     }
 
     public void updateLoanSummaryDerivedFields(final Loan loan) {
+        updateLoanSummaryDerivedFieldsInternal(loan, false);
+    }
+
+    public void updateLoanSummaryDerivedFieldsViaPersistence(final Loan loan) {
+        updateLoanSummaryDerivedFieldsInternal(loan, true);
+    }
+
+    private void updateLoanSummaryDerivedFieldsInternal(final Loan loan, final boolean viaPersistence) {
         flushModeHandler.withFlushMode(FlushModeType.COMMIT, () -> {
             if (loan.isNotDisbursed()) {
                 if (loan.getSummary() != null) {
                     loan.getSummary().zeroFields();
                 }
                 loan.setTotalOverpaid(null);
             } else {
-                refreshSummaryAndBalancesForDisbursedLoan(loan);
+                refreshSummaryAndBalancesForDisbursedLoanInternal(loan, viaPersistence);
             }
         });
     }
 
     public void refreshSummaryAndBalancesForDisbursedLoan(final Loan loan) {
-        final Money overpaidBy = calculateTotalOverpayment(loan);
+        refreshSummaryAndBalancesForDisbursedLoanInternal(loan, false);
+    }
+
+    private void refreshSummaryAndBalancesForDisbursedLoanInternal(final Loan loan, final boolean viaPersistence) {
+        final Money overpaidBy = viaPersistence ? calculateTotalOverpaymentViaPersistence(loan) : calculateTotalOverpayment(loan);
```

Comment:
I dont think we should maintain two versions... use only 1 which fetches the required transactions.